### DISCUSSION TITLE
How to validate query params and apply to getTableQuery() in ListRecords?

### DISCUSSION THREAD

**Discussion by hrvoj3e** [filamentphp/filament#4338] [Help]
Sorry, maybe this is in docs or discussions but I cannot find this.

1. Where should one do validation of query params ?
2. Where should they be stored if validation passed?
3. How to use them in methods like `getTableQuery()`?

```php
class MyList extends ListRecords
{
    protected static string $resource = MyResource::class;
    public ?int $myIntParamInUrl = null;

    public function mount(): void
    {
        parent::mount();

        $validated = collect(Validator::validate(request()->all(), [
            'myIntParamInUrl' => 'integer|min:10000',
        ]));

        if ($t = $validated->get('myIntParamInUrl')) {
             $this->myIntParamInUrl = $t;
        }
    }

    protected function getTableQuery(): Builder
    {
        $builder = parent::getTableQuery();

        if ($this->myIntParamInUrl) {
            $builder->whereHas('myrelation', function($query) {
                $query->where('my_int_param', '=', $this->myIntParamInUrl);
            });
        }

        return $builder;
    }

}
```

Is `mount()` that place? Or directly in `getTableQuery()` ?

The problem right now is that `Validator::validate` throws errors which silently fail (probably Filament catches them).
How should I reneder them to user?

when I use `/?myIntParamInUrl=10001` it is working, but when I use `/?myIntParamInUrl=broken10001` it just shows default list.

**Comment by hrvoj3e** [comment]
I tried to use filter with validation but no validation errors are shown.

Any pointers how to do this?
Am I going in the right direction with this?

```php
    protected function getTableFilters(): array
    {
        $filters = parent::getTableFilters();

        $filters[] = Filters\Filter::make('myIntParamInUrl')
        ->form([
            Forms\Components\TextInput::make('myIntParamInUrl')
                ->filled()
                ->minValue(10000)
                ->numeric()
                // how to populate state from request() ??
        ])
        ->query(function (Builder $query, array $data): Builder {
            return $query->when(
                $data['myIntParamInUrl'],
                fn (Builder $query, $value): Builder => $query->whereHas('myrelation', function ($query) use ($value) {
                    $query->where('my_int_param', '=', $value);
                }),
            );
        })
        ;

        return $filters;
    }


    protected function onValidationError(ValidationException $exception): void
    {
        Notification::make()
            ->title($exception->getMessage())
            ->danger()
            ->send();
    }

```

**Reply by studiowizjo** [reply]
According to what you send me via email:

> What I am trying to do? I must generate a link in another web app to filament admin which must filter resource list based on query parameters.
> 
> I am trying to use validation detect if query param is invalid.
> In my case, param should be numeric and greater than 10000. If it is present in request and is valid it should filter resource list, but if it is invalid it should not show any results.

the direction with [table filters](https://filamentphp.com/docs/2.x/tables/filters) is correct.

Using it you don`t need to create your own variables and validation logic.

You can use `default()` method to populate form field from request:

```php
->default(request('myIntParamInUrl'))
```

**Reply by hrvoj3e** [reply]
Ok, `default()` works.
But why are validation errors via `onValidationError` not shown? Maybe it is not possible from filter forms on ListRecords?

**Reply by studiowizjo** [reply]
I'm not completely sure because I've never used the `onValidationError()` method, but I suppose it's intended to work with `form()` method, not any other forms which you can use on page or in filters.

Try to use `rules()` on your field:

```php
->rules([
    function () {
        return function (string $attribute, $value, Closure $fail) {
            if ($value < 10000) {
                $errorMessage = "{$attribute} cannot be less than 10000.";

                $fail($errorMessage);

                Notification::make()
                    ->title($errorMessage)
                    ->danger()
                    ->send();
            }
        };
    },
])
```

**Comment by hrvoj3e** [comment]
There is another gotcha.

Url param `myIntParamInUrl` and `tableFilters` are present in the same time in url. Changing only `myIntParamInUrl` does not update filtering.
```
# will be OK
&myIntParamInUrl=99999&tableFilters[myIntParamInUrl][myIntParamInUrl]=99999
# will FAIL
&myIntParamInUrl=77777&tableFilters[myIntParamInUrl][myIntParamInUrl]=99999
```

My solution was

```php
  if (request()->has('myIntParamInUrl')) {
      $t = request('myIntParamInUrl');
      $this->tableFilters['myIntParamInUrl']['myIntParamInUrl'] = $t;
      // use livewire directly to remove current param from url
      $this->queryString['myIntParamInUrl'] = ['except' => $t];
  }
```
